### PR TITLE
Redeem beta invites

### PR DESCRIPTION
**Pull request by abhigyanghosh30**
## Done

- Beta participants can redeem their invitation codes

## QA

- Check out this feature branch
- Run the site using the command `./run serve` or `dotrun`
- View the site locally in your web browser at: http://0.0.0.0:8001/
    - Be sure to test on mobile, tablet and desktop screen sizes
- Generate activation key from the back office tool
- Go to /credentials/shop/redeem?code={activationKey}
- It should attach an exam to the user
## Issue / Card

Fixes https://warthogs.atlassian.net/browse/CRED-230

## Screenshots

[If relevant, please include a screenshot.]

## Help

[QA steps](https://discourse.canonical.com/t/qa-steps/152) - [Commit guidelines](https://discourse.canonical.com/t/commit-guidelines/148)

### CODE REVIEW

File: webapp/shop/cube/demo_keys.json [hunk 1]
```diff
@@ -0,0 +1,143 @@
+[
```
Comment (abhigyanghosh30) [inline webapp/shop/cube/demo_keys.json:1]: need to remove this file

File: webapp/shop/api/ua_contracts/api.py [hunk 2]
```diff
@@ -1,4 +1,5 @@
 from typing import Optional
+from urllib import request
```
Comment (abhigyanghosh30) [inline webapp/shop/api/ua_contracts/api.py:2]: remove this too

File: webapp/shop/cube/views.py [hunk 3]
```diff
@@ -710,3 +713,88 @@ def cred_syllabus_data(**kawrgs):
         syllabus_data=syllabus_data,
         exam_name=exam_name,
     )
+
+
+@shop_decorator(area="cube", permission="user", response="html")
+def cred_shop(**kwargs):
```
Comment (mrgnr) [inline webapp/shop/cube/views.py:719]: I'm able to access http://localhost:8001/credentials/shop without being logged in, but this view should only be visible to logged in users.

File: webapp/shop/cube/views.py [hunk 4]
```diff
@@ -710,3 +713,88 @@ def cred_syllabus_data(**kawrgs):
...
+    return flask.render_template("credentials/shop/index.html")
+
+
+@shop_decorator(area="cube", permission="user", response="html")
+def cred_redeem_code(ua_contracts_api, advantage_mapper, **kwargs):
```
Comment (mrgnr) [inline webapp/shop/cube/views.py:724]: http://localhost:8001/credentials/redeem is visible when not logged in and always displays a success notification. This should be behind SSO and show an error if no activation key was provided.

**Review by mrgnr — CHANGES_REQUESTED** [review]
Redeeming keys works well, but the shop and redeem views need to be behind SSO.

File: webapp/shop/cube/views.py (hunk 3 above)
Comment (abhigyanghosh30) [reply webapp/shop/cube/views.py:719]: should be fixed in the latest commit

Comment (abhigyanghosh30) [reply webapp/shop/cube/views.py:720]: @mrgnr can you review the changes and merge this? Has been open for a long time

**Review by mrgnr — APPROVED** [review]
Looks good :+1: I think you need to rebase on `main`, then should be good to merge.

**Comment by linux-in-flux** [comment]
I followed this flow with a sample key and wound up on a reCAPTCHA page with no details or info. While beta participants will have a context from their email, we may need to think about putting some information on this page for people to acclimate.

**Comment by abhigyanghosh30** [comment]
Moved to https://github.com/canonical/ubuntu.com/pull/12496

**Pull request closed without merging**